File: bst/cpp/lb-05-is-bst.cpp
// https://leetcode.com/problems/validate-binary-search-tree
// it will not work when value is INT_MIN/MAX
bool isBst(TreeNode *root, int mn, int mx) {
    if (!root) return true;
    bool curr = mn <= root->val and root->val <= mx;
    return isBst(root->left, mn, root->val - 1) and isBst(root->right, root->val + 1, mx) and curr;
}
// all testcase pass
bool isBst(TreeNode *root, TreeNode *mn = nullptr, TreeNode *mx = nullptr) {
    if (!root) return true;
    if (mn and root->val <= mn->val) return false;
    if (mx and root->val >= mx->val) return false;
    return isBst(root->left, mn, root) and isBst(root->right, root, mx);
}

File: bst/lb-05-is-bst.py
# This method will not work when subtree is bst but left child is > root
# eg. 3 2 5 1 4
def not_working(root):
    if not root or root.left == None and root.right == None:
        return True
    if not root.left:
        return not_working(root.right) and root.right.data > root.data
    if not root.right:
        return not_working(root.left) and root.left.data < root.data

    return not_working(root.left) and not_working(root.right) and root.left.data < root.data < root.right.data


# So all the left child elements should be < root and
# all the right child elements should be > root.
def bst(root, mn, mx):
    def helper(root, mn, mx):
        if not root:
            return True
        # only need to check if root lies in between min and max
        return mn <= root.data <= mx and helper(root.left, mn, root.data-1) and helper(root.right, root.data+1, mx)
    return helper(root, -100005, +100005)
